### PR TITLE
Teach webp lossless decoder to read Tests/LibGfx/test-inputs/simple-vp8l.webp

### PR DESCRIPTION
**Pull request by nico**
This requires implementing most of the rest of the spec.

This adds decoding of the subimages for the predictor and color transforms and for the meta prefix image, and adds code to use them.

It also adds two tweaks to the bitstream decoding for two cases where webp's bitstream is ever so slightly different from deflate's.

Finally, it adds a test.

With this, we can decode not just Tests/LibGfx/test-inputs/simple-vp8l.webp but many other webp files I tried (…well, 5 – namely, the lossless webp files at https://developers.google.com/speed/webp/gallery2).

The only thing still missing after this is the color index transform (plus random bugfixes).

### CODE REVIEW

File: Userland/Libraries/LibGfx/ImageFormats/WebPLoader.cpp [hunk 1]
```diff
@@ -692,6 +692,238 @@ class Transform {
...
+#undef BLUE
+#undef GREEN
+#undef RED
+#undef ALPHA
+    }
+
+    // "Clamp the input value between 0 and 255."
+    static int Clamp(int a)
+    {
+        return (a < 0) ? 0 : (a > 255) ? 255
```
Comment (Hendiadyoin1) [inline Userland/Libraries/LibGfx/ImageFormats/WebPLoader.cpp:793]: We have `AK::clamp(val, min, max)` in StdLibExtras

File: Userland/Libraries/LibGfx/ImageFormats/WebPLoader.cpp [hunk 2]
```diff
@@ -298,19 +298,75 @@ static ErrorOr<VP8LHeader> decode_webp_chunk_VP8L_header(WebPLoadingContext& con
...
+    CanonicalCode()
+        : m_code(0)
+    {
+    }
+
+    static ErrorOr<CanonicalCode> from_bytes(ReadonlyBytes);
+    ErrorOr<u32> read_symbol(LittleEndianInputBitStream&) const;
+
+private:
+    CanonicalCode(u32 single_symbol)
```
Comment (linusg) [inline Userland/Libraries/LibGfx/ImageFormats/WebPLoader.cpp:318]: `explicit`

File: Userland/Libraries/LibGfx/ImageFormats/WebPLoader.cpp [hunk 3]
```diff
@@ -298,19 +298,75 @@ static ErrorOr<VP8LHeader> decode_webp_chunk_VP8L_header(WebPLoadingContext& con
...
+        : m_code(single_symbol)
+    {
+    }
+
+    CanonicalCode(Compress::CanonicalCode code)
```
Comment (linusg) [inline Userland/Libraries/LibGfx/ImageFormats/WebPLoader.cpp:323]: `explicit`

File: Userland/Libraries/LibGfx/ImageFormats/WebPLoader.cpp [hunk 4]
```diff
@@ -692,6 +692,237 @@ class Transform {
...
+// https://developers.google.com/speed/webp/docs/webp_lossless_bitstream_specification#41_predictor_transform
+class PredictorTransform : public Transform {
+public:
+    static ErrorOr<NonnullOwnPtr<PredictorTransform>> read(WebPLoadingContext&, LittleEndianInputBitStream&, IntSize const& image_size);
+    virtual ErrorOr<void> transform(Bitmap&) override;
+
+private:
+    PredictorTransform(int size_bits, NonnullRefPtr<Bitmap> predictor_bitmap)
+        : m_size_bits(size_bits)
+        , m_predictor_bitmap(predictor_bitmap)
```
Comment (linusg) [inline Userland/Libraries/LibGfx/ImageFormats/WebPLoader.cpp:748]: `move()` (yes, that's a micro-optimization :))

File: Userland/Libraries/LibGfx/ImageFormats/WebPLoader.cpp [hunk 5]
```diff
@@ -692,6 +692,237 @@ class Transform {
...
+            .value();
+    }
+    }
+    return Error::from_string_literal("WebPImageDecoderPlugin: invalid predictor");
+}
+
+ARGB32 PredictorTransform::inverse_transform(ARGB32 pixel, ARGB32 prediction)
+{
+    Color pixel_color = Color::from_argb(pixel);
+    Color prediction_color = Color::from_argb(prediction);
```
Comment (linusg) [inline Userland/Libraries/LibGfx/ImageFormats/WebPLoader.cpp:962]: `auto`

File: Userland/Libraries/LibGfx/ImageFormats/WebPLoader.cpp [hunk 6]
```diff
@@ -923,6 +923,85 @@ ARGB32 PredictorTransform::inverse_transform(ARGB32 pixel, ARGB32 prediction)
...
+    // https://developers.google.com/speed/webp/docs/webp_lossless_bitstream_specification#51_roles_of_image_data
+    // "Each ColorTransformElement 'cte' is treated as a pixel whose alpha component is 255,
+    // red component is cte.red_to_blue, green component is cte.green_to_blue
+    // and blue component is cte.green_to_red."
+    Color transform_color = Color::from_argb(transform);
+    i8 red_to_blue = static_cast<i8>(transform_color.red());
+    i8 green_to_blue = static_cast<i8>(transform_color.green());
+    i8 green_to_red = static_cast<i8>(transform_color.blue());
+
+    Color pixel_color = Color::from_argb(pixel);
```
Comment (linusg) [inline Userland/Libraries/LibGfx/ImageFormats/WebPLoader.cpp:1034]: `auto`

**Review by linusg — COMMENTED** [review]
couple of nits

**Comment by nico** [comment]
Thanks, all done!

File: Userland/Libraries/LibCompress/Deflate.h [hunk 7]
```diff
@@ -45,8 +45,10 @@ class CanonicalCode {
     size_t m_max_prefixed_code_length { 0 };
 
     // Compression - indexed by symbol
-    Array<u16, 288> m_bit_codes {}; // deflate uses a maximum of 288 symbols (maximum of 32 for distances)
-    Array<u16, 288> m_bit_code_lengths {};
+    // Deflate uses a maximum of 288 symbols (maximum of 32 for distances),
+    // but this is also used by webp, which can use up to 256 + 24 + (1 << 11) == 2328 symbols.
```
Comment (nico) [inline Userland/Libraries/LibCompress/Deflate.h:49]: #24700 finally adds a use of (and a test for) this.

**Pull request merged**